src/components/Form.js

```diff
@@ -0,0 +1,217 @@
+        _.each(_.keys(this.inputRefs), inputID => (
+            this.touchedInputs[inputID] = true
+        ));
+
+        // Validate form and return early if any errors are found
+        const values = this.getValues();
+        if (!_.isEmpty(this.validate(values))) {
+            return;
+        }
+
+        // Set loading state and call submit handler
+        FormActions.setIsSubmitting(this.props.formID, true);
+        this.props.onSubmit(values);
+    }
+
+    /**
+     * @param {Object} values - An object containing the value of each inputID
+     * @returns {Object} - An object containing the errors for each inputID
+     */
+    validate(values) {
+        FormActions.setServerErrorMessage(this.props.formID, '');
+        const validationErrors = this.props.validate(values);
+        const errors = _.pick(validationErrors, (inputValue, inputID) => (
+            Boolean(this.touchedInputs[inputID])
+        ));
+        this.setState({errors});
+        return errors;
+    }
+
+    render() {
+        const childrenWrapperWithProps = children => (
+            // eslint-disable-next-line rulesdir/prefer-underscore-method
+            React.Children.map(children, (child) => {
+                // Just render the child if it is not a valid React element, e.g. text within a <Text> component
+                if (!React.isValidElement(child)) {
+                    return child;
+                }
+
+                // Depth first traversal of the render tree as the input element is likely to be the last node
+                if (child.props.children) {
+                    return React.cloneElement(child, {
+                        children: childrenWrapperWithProps(child.props.children),
+                    });
+                }
+
+                // We check if the child has the isFormInput prop.
+                // We don't want to pass form props to non form components, e.g. View, Text, etc
+                if (!child.props.isFormInput) {
+                    return child;
+                }
+
+                // We clone the child passing down all form props
+                const inputID = child.props.inputID;
+                return React.cloneElement(child, {
+                    ref: node => this.inputRefs[inputID] = node,
+                    defaultValue: this.props.draftValues[inputID] || child.props.defaultValue,
+                    errorText: this.state.errors[inputID] || '',
+                    onBlur: (event) => {
+                        this.setTouchedInput(inputID);
+                        this.validate(this.getValues());
+                        if (child.props.onBlur) {
+                            child.props.onBlur(event);
+                        }
+                    },
+                    onChange: (value) => {
+                        // Expose value to the input ref so we can access it on native too
+                        this.inputRefs[inputID].value = value;
```

Comment:
I agree with your concerns here. However, I actually searched this quite a bit, but couldn't find anything in the React Native docs or elsewhere. It seems that `TextInput` is a [controlled input](https://reactnative.dev/docs/textinput#value) and ref doesn't expose `value` on native. While testing, I confirmed that `ref.value` is exposed on Web, but not on native. Maybe I'm missing something?

Given what I found above, I opted to expose it in the From because every native input will have to expose `value` in the ref in order to work with `Form`. I think this will DRY up code and make refactoring each input easier.

Reply:
Hmm I might actually have found an issue with the current approach. On native, if inputs are populated with draft values and we just focus and blur them, without actually changing a value we would return undefined in `getValues`. Will test this further and if confirmed work on a fix.

Reply:
Oh hmm, maybe there's no way then. I wasn't actually sure just thought we should look into that. Finding a native solution is a NAB then, but I have a few other thoughts on the next comment...

>I think this will DRY up code and make refactoring each input easier.

- It might be less effort, but still feels like we are making the Form more accepting different kinds of inputs and perhaps it would be better for the Form to have a clear interface that all inputs have to match to be compatible. Then we do not need to leave any platform-dependent comments in the code.

- It seems that in overwriting `ref.value` in `onChange()` we will have a platform inconsistency. Sorry if I'm missing something (I might be), but...

1. Web has `inputRef.value` and native does not
2. If we pass a `defaultValue` to an uncontrolled component then `inputRef.value` on web will be that value instantly. On native the value would be `undefined` until the `onChange` event is triggered.

Thoughts?

Reply:
Sorry that comment got posted before I refreshed looks like you picked up what I was putting down.

Reply:
That makes sense! I mentioned 2 in my comment above and moving this to each input will likely solve the issue. I'll work on that next. Thank you!

Reply:
I moved the ref exposure to the TextInput component on native only. Let me know what you think of the changes.

Reply:
Resolving this conversation in favor of the more recent comments.